mlrun/runtimes/function.py

```diff
@@ -1396,3 +1318,130 @@ def get_nuclio_deploy_status(
+
+    # resolve work_dir and handler
+    work_dir, handler = _resolve_work_dir_and_handler(function.spec.function_handler)
+    work_dir = function.spec.workdir or work_dir
+    if work_dir != "":
+        code_entry_attributes["workDir"] = work_dir
+
+    # archive
+    if code_entry_type == "archive":
+        if source.startswith("v3io"):
+            source = f"http{source[len('v3io'):]}"
+
+        secrets = get_secrets(["V3IO_ACCESS_KEY"])
+        v3io_access_key = secrets.get("V3IO_ACCESS_KEY", "")
+        if v3io_access_key:
+            code_entry_attributes["headers"] = {
+                "headers": {"X-V3io-Session-Key": v3io_access_key}
+            }
+
+    # s3
+    if code_entry_type == "s3":
+        bucket, item_key = parse_s3_bucket_and_key(source)
+
+        code_entry_attributes["s3Bucket"] = bucket
+        code_entry_attributes["s3ItemKey"] = item_key
+
+        secrets = get_secrets(
+            ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"]
+        )
+        code_entry_attributes["s3AccessKeyId"] = secrets.get("AWS_ACCESS_KEY_ID", "")
+        code_entry_attributes["s3SecretAccessKey"] = secrets.get(
+            "AWS_SECRET_ACCESS_KEY", ""
+        )
+        code_entry_attributes["s3SessionToken"] = secrets.get("AWS_SESSION_TOKEN", "")
+
+    # git
+    if code_entry_type == "git":
+
+        # change git:// to https:// as nuclio expects it to be
+        if source.startswith("git://"):
+            source = source.replace("git://", "https://")
+
+        source, reference, branch = _resolve_git_reference_from_source(source)
+        if not branch and not reference:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "git branch or refs must be specified in the source e.g.: "
+                "'git://<url>/org/repo.git#<branch-name or refs/heads/..>'"
+            )
+        if reference:
+            code_entry_attributes["reference"] = reference
+        if branch:
+            code_entry_attributes["branch"] = branch
+
+        secrets = get_secrets(["GIT_USERNAME", "GIT_PASSWORD"])
+        code_entry_attributes["username"] = secrets.get("GIT_USERNAME", "")
+        code_entry_attributes["password"] = secrets.get("GIT_PASSWORD", "")
+
+    # populate spec with relevant fields
+    spec.set_config("spec.handler", handler)
+    spec.set_config("spec.build.path", source)
+    spec.set_config("spec.build.codeEntryType", code_entry_type)
+    spec.set_config("spec.build.codeEntryAttributes", code_entry_attributes)
+
+
+def _resolve_git_reference_from_source(source):
+    # kaniko allow multiple "#" e.g. #refs/..#commit
+    split_source = source.split("#", 1)
+
+    # no reference was passed
+    if len(split_source) < 2:
+        return source, "", ""
+
+    reference = split_source[1]
+    if reference.startswith("refs/"):
+        return split_source[0], reference, ""
+
+    return split_source[0], "", reference
+
+
+def _resolve_work_dir_and_handler(handler):
```

Comment:
same

Reply:
same, spec need to have client side logic or cross functions (base) logic not deploy/crd logic

Reply:
The same I meant here was to add specific unit tests for this function
But yeah, also same for stop using those endless out of context global functions